core_mmu: fix implicit behavior of core_mmu_add_mapping()

a-rybakov opened the pull request:
In core_mmu_add_mapping() requested physical address rounded up/down to granule size (0x100000), which leads
to establising of virtual mappings with overlapped physical counterparts. If two virtual mappings overlaps due to such roundings, then following phys_to_virt() can implicitly return result of virtual address from unexpected mapping. This patch fix such behavior by returning virtual address of newly established mapping.

jenswi-linaro commented on core/drivers/crypto/caam/hal/common/hal_cfg.c:
```diff
@@ -31,19 +31,14 @@ enum caam_status caam_hal_cfg_get_conf(struct caam_jrcfg *jrcfg)
 	if (!ctrl_base) {
 		ctrl_base = (vaddr_t)phys_to_virt(CAAM_BASE, MEM_AREA_IO_SEC);
```
Could do `core_mmu_add_mapping()` hear  and remove some of the code below instead.

jenswi-linaro commented on core/arch/arm/mm/core_mmu.c:
```diff
@@ -1937,39 +1937,39 @@ bool core_mmu_add_mapping(enum teecore_memtypes type, paddr_t addr, size_t len)
 	map = find_map_by_type_and_pa(type, addr);
 	if (map && pbuf_inside_map_area(addr, len, map))
-		return true;
+		return (void *)(vaddr_t)(map->va + (addr - map->pa));
```
The inner `()` are not needed.

jenswi-linaro commented on core/arch/arm/mm/core_mmu.c:
```diff
@@ -2001,7 +2001,7 @@ bool core_mmu_add_mapping(enum teecore_memtypes type, paddr_t addr, size_t len)
 	dsb_ishst();
 
-	return true;
+	return (void *)(vaddr_t)(map->va + (addr - map->pa));
```
The inner `()` are not needed.

jenswi-linaro reviewed (commented):
Nice catch! A few comments, with those addressed:
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

jenswi-linaro commented:
Looks good, please apply:
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

jenswi-linaro commented on core/arch/arm/mm/core_mmu.c:
```diff
@@ -2320,7 +2323,7 @@ vaddr_t io_pa_or_va(struct io_pa_va *p)
 	if (cpu_mmu_enabled()) {
 		if (!p->va)
-			p->va = (vaddr_t)phys_to_virt_io(p->pa);
+			p->va = (vaddr_t)phys_to_virt_io(p->pa, UNSAFE_LEN);
```
`io_pa_or_va()` should also take a `len` parameter.

jenswi-linaro commented on core/arch/arm/mm/core_mmu.c:
```diff
@@ -2332,7 +2335,7 @@ vaddr_t io_pa_or_va_secure(struct io_pa_va *p)
 	if (cpu_mmu_enabled()) {
 		if (!p->va)
-			p->va = (vaddr_t)phys_to_virt(p->pa, MEM_AREA_IO_SEC);
+			p->va = (vaddr_t)phys_to_virt(p->pa, MEM_AREA_IO_SEC, UNSAFE_LEN);
```
`io_pa_or_va_secure()` should also take a `len` parameter.

jenswi-linaro commented on core/arch/arm/mm/core_mmu.c:
```diff
@@ -2344,7 +2347,7 @@ vaddr_t io_pa_or_va_nsec(struct io_pa_va *p)
 	if (cpu_mmu_enabled()) {
 		if (!p->va)
-			p->va = (vaddr_t)phys_to_virt(p->pa, MEM_AREA_IO_NSEC);
+			p->va = (vaddr_t)phys_to_virt(p->pa, MEM_AREA_IO_NSEC, UNSAFE_LEN);
```
`io_pa_or_va_nsec()` should also take `len` parameter.

jenswi-linaro commented on core/arch/arm/mm/core_mmu_lpae.c:
```diff
@@ -700,11 +700,11 @@ bool core_mmu_find_table(struct mmu_partition *prtn, vaddr_t va,
 #ifdef CFG_VIRTUALIZATION
 		if (prtn == &default_partition)
-			tbl = phys_to_virt(ntbl, MEM_AREA_TEE_RAM_RW_DATA);
+			tbl = phys_to_virt(ntbl, MEM_AREA_TEE_RAM_RW_DATA, UNSAFE_LEN);
```
Length should `XLAT_TABLE_SIZE`, same below.

jenswi-linaro commented on core/arch/arm/mm/core_mmu_v7.c:
```diff
@@ -516,7 +516,7 @@ bool core_mmu_find_table(struct mmu_partition *prtn, vaddr_t va,
 	} else {
 		paddr_t ntbl = tbl[n] & ~((1 << 10) - 1);
-		void *l2tbl = phys_to_virt(ntbl, MEM_AREA_TEE_RAM_RW_DATA);
+		void *l2tbl = phys_to_virt(ntbl, MEM_AREA_TEE_RAM_RW_DATA, UNSAFE_LEN);
```
Length should be `L2_TBL_SIZE`.

jenswi-linaro commented on core/arch/arm/mm/tee_pager.c:
```diff
@@ -224,7 +224,7 @@ static void pager_unlock(uint32_t exceptions)
 }
 
-void *tee_pager_phys_to_virt(paddr_t pa)
+void *tee_pager_phys_to_virt(paddr_t pa, size_t len __unused)
```
I believe it would be more correct to check that each page covered by `pa .. pa + len - 1` is mapped since the caller should be able to depend on `phys_to_virt()` to be accurate and not just best effort.

A simplified version might be to fail for all ranges spanning more than one page. If that works I think that should be preferable.

jenswi-linaro commented on core/arch/arm/plat-imx/pm/pm-imx7.c:
```diff
@@ -100,14 +100,16 @@ int pm_imx7_iram_tbl_init(void)
 	iram_tbl_virt_addr = phys_to_virt(iram_tbl_phys_addr,
-					  MEM_AREA_TEE_COHERENT);
+					  MEM_AREA_TEE_COHERENT,
+					  UNSAFE_LEN);
```
It looks like the expected length is 16 KiB.

jenswi-linaro commented on core/arch/arm/plat-ls/main.c:
```diff
@@ -186,9 +186,9 @@ void main_init_gic(void)
 
 	gicc_base = (vaddr_t)phys_to_virt(gic_base + gicc_offset,
-					  MEM_AREA_IO_SEC);
+					  MEM_AREA_IO_SEC, UNSAFE_LEN);
```
`GICC_SIZE`?

jenswi-linaro commented on core/arch/arm/plat-ls/main.c:
```diff
@@ -186,9 +186,9 @@ void main_init_gic(void)
+					  MEM_AREA_IO_SEC, UNSAFE_LEN);
 	gicd_base = (vaddr_t)phys_to_virt(gic_base + gicd_offset,
-					  MEM_AREA_IO_SEC);
+					  MEM_AREA_IO_SEC, UNSAFE_LEN);
```
`GICD_SIZE`?
Please fix the other GIC translations too.

a-rybakov replied:
UNSAFE_LEN was born in answer for GIC translations, as attempt to point platform maintainers to define correct GIC sizes. I didnt dared do that by myself because for various platforms gic sizes are varies. For example, there is few platforms, where gic sizes are defined:
at ./core/arch/arm/plat-stm32mp1/platform_config.h:
#define GIC_SIZE			0x2000
#define GICC_OFFSET			0x1000
#define GICD_OFFSET			0x0000

at ./core/arch/arm/plat-rockchip/platform_config.h:

#elif defined(PLATFORM_FLAVOR_rk3399)
#define GIC_BASE		(MMIO_BASE + 0x06E00000)
#define GIC_SIZE		SIZE_M(2)
#define GICD_BASE		GIC_BASE
#define GICR_BASE		(GIC_BASE + SIZE_M(1))

elif defined(PLATFORM_FLAVOR_px30)
#define GIC_BASE		0xff130000
#define GIC_SIZE		SIZE_K(64)
#define GICD_BASE		(GIC_BASE + 0x1000)
#define GICC_BASE		(GIC_BASE + 0x2000)

And so on. So I think that correct sizes for GICs have to be set by platform maintainers in basis on knowledge of  hardware implementation.

jenswi-linaro replied:
Fair enough. For GIC register mappings give the len `1` in this patch. We can follow up with another PR taking care of the obvious cases. That way we can also avoid needing to wait for many acks.

jenswi-linaro reviewed (commented):
I found quite a few places where the needed length wasn't too hard to figure out. Please try to fix the remaining ones. I expect that there will be only very few places left where we'd need to supply a `1` length instead.

a-rybakov commented on core/arch/arm/mm/core_mmu.c:
```diff
@@ -2289,7 +2292,7 @@ bool cpu_mmu_enabled(void)
 vaddr_t core_mmu_get_va(paddr_t pa, enum teecore_memtypes type)
```
It seems that size can also be determined in a sufficiently large number of places with core_mmu_get_va(). I will try to add length parameter for it too.

jenswi-linaro commented on core/arch/arm/plat-stm32mp1/drivers/stm32mp1_pwr.c:
```diff
@@ -45,7 +45,7 @@ vaddr_t stm32_pwr_base(void)
 	static struct io_pa_va base = { .pa = PWR_BASE };
 
-	return io_pa_or_va_secure(&base);
+	return io_pa_or_va_secure(&base, 0x1000);
```
Please use `1` instead this magic number.

jenswi-linaro commented on core/arch/arm/plat-stm32mp1/drivers/stm32mp1_rcc.c:
```diff
@@ -20,7 +20,7 @@ vaddr_t stm32_rcc_base(void)
 	static struct io_pa_va base = { .pa = RCC_BASE };
 
-	return io_pa_or_va_secure(&base);
+	return io_pa_or_va_secure(&base, 0x1000);
```
Please use `1` instead of this magic number.

jenswi-linaro commented on core/include/mm/core_memprot.h:
```diff
@@ -50,6 +50,12 @@ enum buf_is_attr {
+ * Third parameter for phys_to_virt() and phys_to_virt_io() in case of
+ * the requested length cannot be specified or doesn`t matter
+ */
+#define UNSAFE_LEN	0
```
I guess we can remove this define now.

jenswi-linaro commented on core/pta/bcm/elog.c:
```diff
@@ -89,12 +89,12 @@ static TEE_Result pta_elog_load_nitro_fw(uint32_t param_types,
 	sz = params[0].memref.size;
 
+	src_vaddr = (vaddr_t)phys_to_virt((uintptr_t)src_paddr + offset,
+					  MEM_AREA_RAM_SEC, sz);
```
If we make the effort to figure out the size of the memory area then I we should also check the return value of this function and not just crash in `memcpy()` if this fails.

jenswi-linaro commented on core/pta/bcm/elog.c:
```diff
@@ -155,8 +156,10 @@ static TEE_Result pta_elog_nitro_crash_dump(uint32_t param_types,
+			DEFAULT_ELOG_BUFFER_SIZE : params[0].memref.size;
 	src_vaddr = (vaddr_t)phys_to_virt((uintptr_t)src_paddr + offset,
-					  MEM_AREA_RAM_SEC);
+					  MEM_AREA_RAM_SEC, sz);
```
We should probably feed `DEFAULT_ELOG_BUFFER_SIZE` instead of `sz` here. Please check the return value also.

a-rybakov replied:
Seems that sz should be used in cache_op_inner() below...

jenswi-linaro replied:
Possibly, but I'd say that's out of scope for this PR.

a-rybakov replied:
Is this means that it is better to leave cache_op_inner() as it was before?

jenswi-linaro replied:
Yes

jenswi-linaro commented on core/pta/bcm/elog.c:
```diff
@@ -179,13 +182,16 @@ static TEE_Result pta_elog_dump(uint32_t param_types,
-	src_vaddr = (vaddr_t)phys_to_virt(src_paddr, MEM_AREA_RAM_NSEC);
+	sz = (params[0].memref.size > DEFAULT_ELOG_BUFFER_SIZE) ?
+			DEFAULT_ELOG_BUFFER_SIZE : params[0].memref.size;
+	src_vaddr = (vaddr_t)phys_to_virt(src_paddr, MEM_AREA_RAM_NSEC, sz);
```
Please check the return value.

jenswi-linaro commented on core/pta/bcm/elog.c:
```diff
@@ -155,8 +156,10 @@ static TEE_Result pta_elog_nitro_crash_dump(uint32_t param_types,
 		return TEE_ERROR_ACCESS_DENIED;
 	}
 
+	sz = (params[0].memref.size > DEFAULT_ELOG_BUFFER_SIZE) ?
```
Please use the `MIN()` macro instead.

jenswi-linaro commented on core/pta/bcm/elog.c:
```diff
@@ -179,13 +182,16 @@ static TEE_Result pta_elog_dump(uint32_t param_types,
 	}
 
-	src_vaddr = (vaddr_t)phys_to_virt(src_paddr, MEM_AREA_RAM_NSEC);
+	sz = (params[0].memref.size > DEFAULT_ELOG_BUFFER_SIZE) ?
```
Please use the `MIN()` macro instead.

jenswi-linaro reviewed (commented):
Comments for the commit "core_mmu: fix implicit behavior of core_mmu_add_mapping()"

When you update, please do an incremental update instead. It will save some time for the reviewers  and is the preferred way to address comments. Sometimes it might make more sense to do a force push, but most often not. If you're not certain, please ask.

a-rybakov commented on core/arch/arm/mm/core_mmu.c:
```diff
@@ -194,11 +194,16 @@ static bool _pbuf_is_inside(struct memaccess_area *a, size_t alen,
-static bool pa_is_in_map(struct tee_mmap_region *map, paddr_t pa)
+static bool pa_is_in_map(struct tee_mmap_region *map, paddr_t pa, size_t len)
 {
-	if (!map)
```
Nice suggestion! Thanks)

a-rybakov commented on core/arch/arm/plat-imx/main.c:
```diff
@@ -120,7 +120,7 @@ void main_init_gic(void)
 	vaddr_t gicd_base;
 
-	gicd_base = core_mmu_get_va(GICD_BASE, MEM_AREA_IO_SEC);
+	gicd_base = core_mmu_get_va(GICD_BASE, MEM_AREA_IO_SEC, 0x10000);
```
It can be determined as 0x10000 about 50 lines above, but ok, it seems as magic number here.

a-rybakov commented:
> Comments for the commit "core_mmu: fix implicit behavior of core_mmu_add_mapping()"
> 
> When you update, please do an incremental update instead. It will save some time for the reviewers and is the preferred way to address comments. Sometimes it might make more sense to do a force push, but most often not. If you're not certain, please ask.

Is it correct to perform force-push to fix small mistake in commit with portion of addressed comments?

jenswi-linaro commented:
> > Comments for the commit "core_mmu: fix implicit behavior of core_mmu_add_mapping()"
> > When you update, please do an incremental update instead. It will save some time for the reviewers and is the preferred way to address comments. Sometimes it might make more sense to do a force push, but most often not. If you're not certain, please ask.
> 
> Is it correct to perform force-push to fix small mistake in commit with portion of addressed comments?

It's better to do an incremental update that can be squashed up in the end before merging. Here in the PR we can see that there has been a force push, but often the information on what was replaced is lost or hard to come by.

etienne-lms commented on core/arch/arm/mm/core_mmu.c:
```diff
@@ -2213,11 +2214,17 @@ static void *phys_to_virt_ta_vaspace(paddr_t pa, size_t len)
 #ifdef CFG_WITH_PAGER
 static void *phys_to_virt_tee_ram(paddr_t pa, size_t len)
 {
+	paddr_t pa_end;
```
`paddr_t pa_end = 0;`

and s/end_pa/pa_end/g at line 2219.

etienne-lms commented on core/arch/arm/plat-imx/mmdc.c:
```diff
@@ -17,17 +17,11 @@
-
-	mmdc_base = core_mmu_get_va(MMDC_P0_BASE, MEM_AREA_IO_SEC,
-				    off + sizeof(uint32_t));
+	uint32_t val;
```
`	uint32_t val = 0;`

a-rybakov replied:
Now val and off are left as they were before my commit, so I wondering - is it ok now?

jenswi-linaro replied:
A few years ago we changed the coding guidelines to always initialize all local variables with some value. Not all functions follow this convention. When we make changes to such functions we often take care of this initialization too, especially if we make changes to local variables. So please initialize all the currently uninitialized locals with `0`, preferably with not more than one variable per line.

jenswi-linaro commented on core/arch/arm/plat-imx/mmdc.c:
```diff
@@ -17,17 +17,11 @@
-	mmdc_base = core_mmu_get_va(MMDC_P0_BASE, MEM_AREA_IO_SEC,
-				    off + sizeof(uint32_t));
+	uint32_t val;
+	uint32_t off = is_mx7 ? DDRC_MSTR : MMDC_MDMISC;
```
Please leave the if as it was.

a-rybakov commented:
Hello!

How do you think, is it ok to add some 'debug' patches? The whole patchset fails on hikey-6220 platform, so I want to do some debugging for it.

jforissier commented:
Don't worry about the message `Panic at core/mm/vm.c:735 <vm_set_prot>`, it is a known issue, a fix is available and pending review (#4628).

However, the Azure CI must pass. Please rebase onto current master.

jenswi-linaro commented:
Hi!

I'm sorry, It seems I made this harder for you than necessary.

Unless you can get it sorted soon. How about supplying the size `1` all the calls to `phys_to_virt()` and friends (more or less what you proposed originally if I remember correctly). That should have minimal impact on any platform and all tests should pass so we can merge this.

After that you could create other PRs with updated size parameter for `phys_to_virt()` at different locations (based  on the diffs you already have here).

Just a though, feel free to continue in the current direction if you think that's easier.

etienne-lms commented on core/arch/arm/kernel/virtualization.c:
```diff
@@ -225,8 +226,9 @@ static int configure_guest_prtn_mem(struct guest_partition *prtn)
-	       __data_end - __data_start);
+	       phys_to_virt(original_data_pa, MEM_AREA_SEC_RAM_OVERALL,
+			    __data_end - __data_start),
+			    __data_end - __data_start);
```
bad indentation at line 231

etienne-lms commented on core/arch/arm/mm/core_mmu.c:
```diff
@@ -2197,78 +2203,87 @@ static void check_va_matches_pa(paddr_t pa __unused, void *va __unused)
+	paddr_t end_pa = 0;
+
+	if (SUB_OVERFLOW(len, 1, &end_pa) || ADD_OVERFLOW(pa, end_pa, &end_pa))
+		return false;
```
s/false/NULL/

etienne-lms commented on core/arch/arm/mm/tee_pager.c:
```diff
@@ -233,6 +233,9 @@ void *tee_pager_phys_to_virt(paddr_t pa)
 	vaddr_t v;
 	size_t n;
 
+	if (pa & CORE_MMU_PGDIR_MASK || len > CORE_MMU_PGDIR_SIZE)
```
Should be `SMALL_PAGE_MASK` and `SMALL_PAGE_SIZE`, no?

a-rybakov replied:
Yes, now PGDIR looks strange for me too)

etienne-lms commented on core/arch/arm/plat-stm32mp1/main.c:
```diff
@@ -271,12 +271,14 @@ vaddr_t stm32_get_gpio_bank_base(unsigned int bank)
-	return io_pa_or_va_nsec(&gpios_nsec_base) + (bank * GPIO_BANK_OFFSET);
+	return io_pa_or_va_nsec(&gpios_nsec_base,
+				bank * (GPIO_BANK_OFFSET + 1)) +
+				(bank * GPIO_BANK_OFFSET);
```
i think it should be:
```
	return io_pa_or_va_nsec(&gpios_nsec_base,
				(bank + 1) * GPIO_BANK_OFFSET) +
	       bank * GPIO_BANK_OFFSET;

etienne-lms commented on core/arch/arm/plat-vexpress/main.c:
```diff
@@ -52,9 +52,9 @@ void main_init_gic(void)
 
 	gicc_base = (vaddr_t)phys_to_virt(GIC_BASE + GICC_OFFSET,
-					  MEM_AREA_IO_SEC);
+					  MEM_AREA_IO_SEC, GIC_DIST_REG_SIZE);
```
s/GIC_DIST_REG_SIZE/GIC_CPU_REG_SIZE/

a-rybakov replied:
It is mapped as GIC_DIST_REG_SIZE, but ok, suppose it can be fixed later.

etienne-lms commented on core/arch/arm/plat-zynq7k/main.c:
```diff
@@ -183,7 +183,8 @@ static uint32_t write_slcr(uint32_t addr, uint32_t val)
 				va = (vaddr_t)phys_to_virt(SLCR_BASE,
-							   MEM_AREA_IO_SEC);
+							   MEM_AREA_IO_SEC,
+							   sizeof(uint32_t));
```
should be
```
phys_to_virt(SLCR_BASE,
             MEM_AREA_IO_SEC 
             addr + sizeof(uint32_t));

Ditto below

etienne-lms commented on core/drivers/bcm_gpio.c:
```diff
@@ -161,7 +161,9 @@ void iproc_gpio_set_secure(int gpiopin)
-						 MEM_AREA_IO_SEC);
+						 MEM_AREA_IO_SEC,
+						 IPROC_GPIO_SEC_CFG_REG(gpiopin)
+						 + sizeof(uint32_t));
```
`+` should be at end of the previous line.
I suggest:
```
	vaddr_t baseaddr =
		(vaddr_t)phys_to_virt(CHIP_SECURE_GPIO_CONTROL0_BASE,
				      MEM_AREA_IO_SEC,
				      IPROC_GPIO_SEC_CFG_REG(gpiopin) +
				      sizeof(uint32_t));
```

etienne-lms commented on core/drivers/hi16xx_rng.c:
```diff
@@ -74,7 +76,9 @@ uint8_t hw_get_random_byte(void)
-		r = (vaddr_t)phys_to_virt(RNG_BASE, MEM_AREA_IO_SEC) + RNG_NUM;
+		r = (vaddr_t)phys_to_virt(RNG_BASE, MEM_AREA_IO_SEC,
+					  RNG_NUM + sizeof(uint32_t))
+					  + RNG_NUM;
```
```
		r = (vaddr_t)phys_to_virt(RNG_BASE, MEM_AREA_IO_SEC,
					  RNG_NUM + sizeof(uint32_t)) +
		    RNG_NUM;

etienne-lms commented on core/drivers/imx_uart.c:
```diff
@@ -35,21 +35,22 @@
+#define UBIR      0xa4 /* BRM Incremental Register */
+#define UBMR      0xa8 /* BRM Modulator Register */
+#define UBRC      0xac /* Baud Rate Count Register */
+#define UTS       0xb4 /* UART Test Register (mx31) */
```
I would not have changed these lines

etienne-lms commented on core/include/mm/core_memprot.h:
```diff
@@ -67,17 +67,23 @@ bool core_pbuf_is(uint32_t flags, paddr_t pbuf, size_t len);
+ * Translate physical address to virtual address using specified mapping.
+ * Also tries to find proper mapping which have counterpart translation
+ * for specified length of data starting from given physical address.
+ * If length doesn`t matter or cannot be precisely defined, then len
```
`cannot be precisely defined` and `hard to specify` (below) sound strange.
Should one set the `len` of not? 
Maybe rephrase to something like: 'Can be 1 if caller knows the area does not spread over mapping granule'.

a-rybakov replied:
You`re right, rephrased this.

etienne-lms commented on core/include/mm/core_memprot.h:
```diff
@@ -67,17 +67,23 @@ bool core_pbuf_is(uint32_t flags, paddr_t pbuf, size_t len);
- * first then MEM_AREA_IO_NSEC if not found.
+ * first then MEM_AREA_IO_NSEC if not found. Like phys_to_virt() tries
+ * to find proper mapping rely on length parameter. If length is hard
+ * to specify or doesn`t matter, then it should be set to UNSAFE_LEN.
```
remove `UNSAFE_LEN`

etienne-lms reviewed (commented):
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` for commit "core_mmu: fix implicit behavior of core_mmu_add_mapping()".

Comments for commit "core_mmu: fix implicit behavior of core_mmu_add_mapping()".

etienne-lms commented on core/include/mm/core_memprot.h:
```diff
@@ -67,17 +67,22 @@ bool core_pbuf_is(uint32_t flags, paddr_t pbuf, size_t len);
  * Translate physical address to virtual address trying MEM_AREA_IO_SEC
- * first then MEM_AREA_IO_NSEC if not found.
+ * first then MEM_AREA_IO_NSEC if not found. Like phys_to_virt() tries
+ * to find proper mapping rely on length parameter.
```
s/rely/relying/

etienne-lms commented on core/include/mm/core_memprot.h:
```diff
@@ -67,17 +67,22 @@ bool core_pbuf_is(uint32_t flags, paddr_t pbuf, size_t len);
+ * Also tries to find proper mapping which have counterpart translation
+ * for specified length of data starting from given physical address.
+ * Len parameter can be set to 1 if caller knows that requested (pa + len)
+ * doesn`t cross granule boundary.
```
suggestions:
 ...for specified **byte** length of ....
 ...doesn`t cross **mapping** granule boundary.

a-rybakov replied:
byte seems exceeding for me)

etienne-lms commented on core/pta/tests/invoke.c:
```diff
@@ -277,8 +277,8 @@ static TEE_Result test_inject_sdp(uint32_t type, TEE_Param p[TEE_NUM_PARAMS])
-	if (test_v2p2v(src) || test_v2p2v(src + sz - 1) ||
-	    test_v2p2v(dst) || test_v2p2v(dst + sz - 1))
+	if (test_v2p2v(src, sz) || test_v2p2v(src + sz - 1, 0) ||
+	    test_v2p2v(dst, sz) || test_v2p2v(dst + sz - 1, 0))
```
passing `0` size is strange, and actually, the 1st test already does both jobs.
What about `if (test_v2p2v(src, sz) || test_v2p2v(dst, sz))` ?

a-rybakov replied:
Looks like here must be passed 1 instead of 0 - to check last byte in range

etienne-lms replied:
Old `test_v2p(va)` only checked 1 single memory cell address conversion, so we used check both start and end addresses of the buffer.

Since now `test_v2p2v(va, sz)` checks base address + buffer consistency in virtual space, i don't think it is necessary to test the address of the buffer's last memory cell, hence only 1 call to `test_v2p()`.

a-rybakov replied:
You`re right, first check covers full range. Fixed this.

etienne-lms commented on core/arch/arm/plat-synquacer/main.c:
```diff
@@ -45,7 +45,8 @@ void main_init_gic(void)
 	gicd_base = (vaddr_t)phys_to_virt(GIC_BASE + GICD_OFFSET,
-					  MEM_AREA_IO_SEC);
+					  MEM_AREA_IO_SEC,
+					  1);
```
maybe move to previous line?

etienne-lms commented on core/drivers/stm32_rng.c:
```diff
@@ -124,7 +124,8 @@ TEE_Result stm32_rng_read_raw(vaddr_t rng_base, uint8_t *out, size_t *size)
 {
-	vaddr_t rng_cr = io_pa_or_va(&dev->base) + RNG_CR;
+	vaddr_t rng_cr = io_pa_or_va(&dev->base, RNG_CR + sizeof(uint32_t)) +
+				     RNG_CR;
```
fix indentation:
```c
	vaddr_t rng_cr = io_pa_or_va(&dev->base, RNG_CR + sizeof(uint32_t)) +
			 RNG_CR;
```
I would prefer here (for the same reason as below):
```c
	vaddr_t rng_cr = io_pa_or_va(&dev->base, 1) + RNG_CR;

etienne-lms commented on core/drivers/stm32_rng.c:
```diff
@@ -149,7 +150,7 @@ TEE_Result stm32_rng_read(uint8_t *out, size_t size)
 	TEE_Result rc = 0;
 	uint32_t exceptions = 0;
-	vaddr_t rng_base = io_pa_or_va(&stm32_rng->base);
+	vaddr_t rng_base = io_pa_or_va(&stm32_rng->base, size);
```
size to fix: it's not the right size to check.
I would prefer here `1` as RNG IOmem interface is known to not cross page boundaries.
```
	vaddr_t rng_base = io_pa_or_va(&stm32_rng->base, 1);
```

a-rybakov replied:
Ok, so I suppose that in drivers/hi16xx_rng.c size in phys_to_virt() should be changed to 1 too.

etienne-lms reviewed (commented):
For commit "core_mmu: fix implicit behavior of core_mmu_add_mapping()":
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>`
I think this commit can be merge as is. @a-rybakov maybe you can create a specific P-R for this change.

Review comments for commit "core_mmu: fix phys_to_virt() to check length".

a-rybakov commented:
> For commit "core_mmu: fix implicit behavior of core_mmu_add_mapping()":
> `Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>`
> I think this commit can be merge as is. @a-rybakov maybe you can create a specific P-R for this change.
> 
> Review comments for commit "core_mmu: fix phys_to_virt() to check length".

Hi! I`ve created new P-R for it - https://github.com/OP-TEE/optee_os/pull/4671

jenswi-linaro commented:
> Hello! All previous comments are addressed, so will anyone have objection about rebase and squash previous commits?

Sounds good to me.

a-rybakov commented:
Hello!

@jforissier Could you please clarify, what does mean message "Job cancelled!" from  IBART? From the logs it seems that it is something wrong with board firmware and doesn`t related to this PR.

jforissier commented:
@a-rybakov unfortunately IBART doesn't show the OP-TEE secure console when running on QEMU, but there seems to be an issue with this PR indeed. [`make run`](https://optee.readthedocs.io/en/latest/building/devices/qemu.html#qemu-v7) ends up with:
```
E/TC:0 0 Panic 'Failed to register tee ram' at core/mm/mobj.c:658 <mobj_init>
E/TC:0 0 TEE load address @ 0x24965000
E/TC:0 0 Call stack:
E/TC:0 0  0x2496b8dd
```

a-rybakov commented:
> E/TC:0 0 Panic 'Failed to register tee ram' at core/mm/mobj.c:658 <mobj_init>
> E/TC:0 0 TEE load address @ 0x24965000
> E/TC:0 0 Call stack:
> E/TC:0 0  0x2496b8dd
> ```

@jforissier Thank you! Will check it.

a-rybakov commented:
Hello!

I investigated this panic and found, that phys_to_virt() with len parameter is working as it should:
* we have two regions TEE_RAM_RX and TEE_RAM_RW placed in two contiguous mappings
* in function mobj_init() we call mobj_phys_alloc() with params:
	** pa = TEE_RAM_START,
	** size = (VCORE_UNPG_RW_PA + VCORE_UNPG_RW_SZ -	TEE_RAM_START)
* size covers whole range of two regions — [TEE_RAM_RX + TEE_RAM_RW]
* so phys_to_virt() fails expectedly because no such amount of memory in each of these mappings in difference.

Now I wondering how to deal with it. Does mobj() really need to manage TEE_RAM_RX memory? If not, seems that size requested in mobj_phys_alloc() should be corrected to TEE_RAM_RW size.
What do you think?

jenswi-linaro commented:
`thread_get_user_kcode()` and `thread_get_user_kdata()` uses `mobj_tee_ram` to map rx and rw memory respectively.
A possible solution might be to use one `mobj_tee_ram_rx` and one `mobj_tee_ram_rw` instead of the combined `mobj_tee_ram`.
We'll obviously need to use new offsets in `thread_get_user_kcode()` and `thread_get_user_kdata()` relative to the new mobjs.

a-rybakov commented:
> `thread_get_user_kcode()` and `thread_get_user_kdata()` uses `mobj_tee_ram` to map rx and rw memory respectively.
> A possible solution might be to use one `mobj_tee_ram_rx` and one `mobj_tee_ram_rw` instead of the combined `mobj_tee_ram`.
> We'll obviously need to use new offsets in `thread_get_user_kcode()` and `thread_get_user_kdata()` relative to the new mobjs.

Thank you for explanation. I`ve created PR #4701 for split mobj_tee_ram to rx/rw parts.

jenswi-linaro commented:
Please rebase on master to see if the recent azure fixes helps.

a-rybakov commented:
It seems that unfortunately rebase didn`t help. I tested it with QEMUv8 locally - all tests passed ok.
+-----------------------------------------------------
25844 subtests of which 0 failed
93 test cases of which 0 failed
0 test cases were skipped
TEE test application done!
# 
# cat /proc/config.gz | gzip -d|grep "CONFIG_ARM64"
CONFIG_ARM64=y

jforissier commented:
@a-rybakov the problem is with virtualization support enabled: https://dev.azure.com/OPTEE/optee_os/_build/results?buildId=789&view=logs&j=705748f3-7146-5e86-79af-1c0266d20a8c&t=d3458a7d-1a81-5639-8e63-8927d807e53a&l=7150

Try `make arm-tf-clean; make -j10 XEN_BOOT=y run`. This is what I get on the secure console (CFG_TEE_CORE_LOG_LEVEL=2 so only I/E messages): 

```
I/TC:
I/TC: Non-secure external DT found
I/TC: Switching console to device: /pl011@9040000
I/TC: OP-TEE version: 3.13.0-110-g8d8376a0 (gcc version 10.2.1 20201103 (GNU Toolchain for the A-profile Architecture 10.2-2020.11 (arm-10.16))) #2 Tue Jul  6 11:57:25 UTC 2021 aarch64
I/TC: Primary CPU initializing
I/TC: Initializing virtualization support
I/TC: Primary CPU switching to normal world boot
I/TC: Removing guest 0
E/TC:0   virt_guest_destroyed:312 Client with id 0 is not found
I/TC: Added guest 1
E/TC:0 0 Panic 'Failed to register tee ram rw' at core/mm/mobj.c:704 <mobj_init>
E/TC:0 0 TEE load address @ 0xb26a3000
E/TC:0 0 Call stack:
E/TC:0 0  0x00000000b26ab668 print_kernel_stack at optee_os/core/arch/arm/kernel/unwind_arm64.c:80
E/TC:0 0  0x00000000b26b7fa4 __do_panic at optee_os/core/kernel/panic.c:24
E/TC:0 0  0x00000000b26bc8a8 mobj_init at optee_os/core/mm/mobj.c:695
E/TC:0 0  0x00000000b26b96b8 call_initcalls at optee_os/core/kernel/initcall.c:22
E/TC:0 0  0x00000000b26ac018 virt_on_stdcall at optee_os/core/arch/arm/kernel/virtualization.c:367
E/TC:0 0  0x00000000b26a9974 __thread_std_smc_entry at optee_os/core/arch/arm/kernel/thread_optee_smc.c:210
```

a-rybakov commented:
> @a-rybakov the problem is with virtualization support enabled: https://dev.azure.com/OPTEE/optee_os/_build/results?buildId=789&view=logs&j=705748f3-7146-5e86-79af-1c0266d20a8c&t=d3458a7d-1a81-5639-8e63-8927d807e53a&l=7150

Aaah, ok, will check it. Thank you.

a-rybakov commented:
Hi All!

When virtualization enabled, memory map looks like this -
....
D/TC:0   prepare_memory_map:130 T: TEE_RAM_RX       rsz: 00001000, pa: 0e100000, va: adf99000, sz: 0004a000 attr: 1851
D/TC:0   prepare_memory_map:130 T: TEE_RAM_RO       rsz: 00001000, pa: 0e14a000, va: adfe3000, sz: 00016000 attr: 1811
D/TC:0   prepare_memory_map:130 T: NEX_RAM_RW       rsz: 00001000, pa: 0e160000, va: adff9000, sz: 0001a000 attr: 1831
D/TC:0   prepare_memory_map:130 T: TEE_RAM_RW       rsz: 00001000, pa: 0e300000, va: ae013000, sz: 00186000 attr: 1831
D/TC:0   prepare_memory_map:130 T: TA_RAM           rsz: 00001000, pa: 0e486000, va: ae286000, sz: 004ec000 attr: 1831
.....

So RX and RW regions are placed non contiguosly and previous patch for splitting RX/RW mobj`s doesn`t handle this. Please see to PR #4744  - it should fix it finally, I hope.

a-rybakov commented:
Rebased to latest master, all checks passed.

etienne-lms commented on core/drivers/crypto/caam/hal/imx_6_7/hal_clk_mx7.c:
```diff
@@ -11,7 +11,7 @@
 void caam_hal_clk_enable(bool enable)
 {
-	vaddr_t ccm_base = (vaddr_t)phys_to_virt(CCM_BASE, MEM_AREA_IO_SEC);
+	vaddr_t ccm_base = (vaddr_t)phys_to_virt(CCM_BASE, MEM_AREA_IO_SEC, 1);
```
From [imx7-crm.h](https://github.com/OP-TEE/optee_os/blob/3.14.0/core/arch/arm/plat-imx/registers/imx7-crm.h#L32-L36), `CCM_CCGRx_{SET|CLR}()` macros do cross 4kB boundaries.   @cneveux, please cross-check this is fine.

cneveux replied:
@etienne-lms 
You're right but this is not a problem in this case as it's a register set mapped on a section not on a small page.

etienne-lms replied:
it's not really a matter of being mapped by section or small page. See the initial comment for this P-R, and more precisely:Issue explained in more details at #4616.

cneveux replied:
In plat-imx, we use the following code
`register_phys_mem_pgdir(MEM_AREA_IO_SEC, [base address], ROUNDUP([size], CORE_MMU_PGDIR_SIZE));` 
to mapped each HW registers block without using the small page mapping. This create scatter `phys_mem_map` table.
Then the `phys_mem_map` table is mapped at boot time with the `core_init_mmu_map()` and I didn't see call to the `core_mmu_add_mapping()` which rebuild the area size function of the TLB information (which define the MMU page granularity).
Hence, except if I missed something the `register_phys_mem_pgdir()` with a length roundup to a section will result to only level 1 MMU Mapping which are section or supersection. 

Am I correct?

etienne-lms replied:
I think you are correct. If you're confident with the implementation here, then i'm fine.

etienne-lms commented on core/drivers/ns16550.c:
```diff
@@ -49,7 +49,8 @@ static void ns16550_flush(struct serial_chip *chip)
 		container_of(chip, struct ns16550_data, chip);
-	vaddr_t base = io_pa_or_va(&pd->base);
+	vaddr_t base = io_pa_or_va(&pd->base, (UART_LSR << pd->reg_shift) +
+					       pd->io_width);
```
nitpicking: indentation issue, prefer:
```c
	vaddr_t base = io_pa_or_va(&pd->base,
				   (UART_LSR << pd->reg_shift) + pd->io_width);
```

Ditto at line 65.

etienne-lms commented on core/drivers/stm32_bsec.c:
```diff
@@ -153,7 +153,8 @@ static uint32_t otp_bank_offset(uint32_t otp_id)
 {
-	return io_pa_or_va_secure(&bsec_dev.base);
+	return io_pa_or_va_secure(&bsec_dev.base, BSEC_IP_MAGIC_ID_OFF +
+				  sizeof(uint32_t));
```
prefer `return io_pa_or_va_secure(&bsec_dev.base, 1);` here.
BSEC iomem does not cross 4KB boundaries.

etienne-lms commented on core/drivers/stm32_uart.c:
```diff
@@ -48,7 +49,7 @@ static vaddr_t loc_chip_to_base(struct serial_chip *chip)
 	pd = container_of(chip, struct stm32_uart_pdata, chip);
 
-	return io_pa_or_va(&pd->base);
+	return io_pa_or_va(&pd->base, UART_SIZE);
```
`1` would be fine

etienne-lms commented on core/pta/bcm/elog.c:
```diff
@@ -89,12 +89,16 @@ static TEE_Result pta_elog_load_nitro_fw(uint32_t param_types,
+					  MEM_AREA_RAM_SEC, sz);
+	if (!src_vaddr) {
+		EMSG("Not enough memory mapped");
+		return TEE_ERROR_OUT_OF_MEMORY;
```
`TEE_ERROR_OUT_OF_MEMORY` return value could be misinterpreted. Prefer `TEE_ERROR_BAD_PARAMETERS` IMO.

etienne-lms commented on core/pta/bcm/elog.c:
```diff
@@ -155,13 +160,18 @@ static TEE_Result pta_elog_nitro_crash_dump(uint32_t param_types,
+					  MEM_AREA_RAM_SEC, sz);
+	if (!src_vaddr) {
+		EMSG("Not enough memory mapped");
+		return TEE_ERROR_OUT_OF_MEMORY;
```
ditto

a-rybakov commented on core/arch/arm/mm/tee_pager.c:
```diff
@@ -233,6 +233,9 @@ void *tee_pager_phys_to_virt(paddr_t pa)
 	vaddr_t v;
 	size_t n;
 
+	if (pa & SMALL_PAGE_MASK || len > SMALL_PAGE_SIZE)
```
I just wondering, do this function will works fine if incoming _pa_ doesn't page aligned -

`core_mmu_get_entry(&ti, idx, &p, &a);`
`if ((a & TEE_MATTR_VALID_BLOCK) && p == pa)`

seems that _p_, extracted from mmu table, always be page aligned. So, if _pa_ doesn`t, then this phys_to_virt() helper will waste time in vain attemps to resolve _pa_. Then it seems that we should check _pa_ param at function enter, or fix function to be able to find _pa_ with offset from page boundary.

What do you think?

etienne-lms replied:
Indeed you're right. I'm a bit surprised only page aligned addresses are converted inside pageable memory. I guess it's not needed.

Discard my comment and sorry for the confusion.

etienne-lms commented on core/pta/bcm/elog.c:
```diff
@@ -179,13 +189,19 @@ static TEE_Result pta_elog_dump(uint32_t param_types,
+	src_vaddr = (vaddr_t)phys_to_virt(src_paddr, MEM_AREA_RAM_NSEC, sz);
+	if (!src_vaddr) {
+		EMSG("Not enough memory mapped");
+		return TEE_ERROR_OUT_OF_MEMORY;
```
would prefer `TEE_ERROR_BAD_PARAMETERS` here.

etienne-lms reviewed (commented):
`Acked-by: Etienne Carriere <etienne.carriere@linaro.org>` with a minor comment.
The generic code looks all good to me.

jenswi-linaro commented:
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`
I'd appreciate feedback from some platform maintainers too, especially for plat-imx. I assume @etienne-lms is covering plat-stm*.

etienne-lms commented:
As for the tests: `Tested-by: Etienne Carriere <etienne.carriere@linaro.org> (stm32mp1-157C_DK2)`

jforissier commented:
I will merge this in a couple of days or so, just giving some more time for @OP-TEE/plat-imx people to chime in should something be wrong (as Jens mentioned quite a lot of plat-imx* files are touched).

cneveux commented:
> 
> 
> I will merge this in a couple of days or so, just giving some more time for @OP-TEE/plat-imx people to chime in should something be wrong (as Jens mentioned quite a lot of plat-imx* files are touched).

@jforissier, 
@clementfaure ran a regression on i.MX devices and it looks good except 2 tests (1006 and 4013) when CAAM is enabled on some devices. It seems to be not link to this PR but @clementfaure is investigating and acked the PR if OK.
